os/pm/pm_timedStay.c

```diff
@@ -0,0 +1,171 @@
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ************************************************************************/
+
+/************************************************************************
+ * Included Files
+ ************************************************************************/
+
+#include <tinyara/pm/pm.h>
+#include <tinyara/wdog.h>
+
+#include <pm_timer/pm_timer.h>
+
+/************************************************************************
+ * Pre-processor Definitions
+ ************************************************************************/
+
+/************************************************************************
+ * Private Type Declarations
+ ************************************************************************/
+
+/************************************************************************
+ * Private Variables
+ ************************************************************************/
+
+/* This array maps the pid to their respective wdog timer. 
+ * Assumption: when a thread does a timed lock , only the same thread 
+ * can unlock the "timed lock" before it expire */
+static WDOG_ID pid_timerMap[CONFIG_MAX_TASKS];
+
+/************************************************************************
+ * Private Functions
+ ************************************************************************/
+
+static void timer_timeout(int argc, uint32_t pid)
+{
+	/* PM transition will be relaxed here */
+	if (pid_timerMap[pid] != NULL) {
+		pm_relax(PM_IDLE_DOMAIN, PM_NORMAL);
+		wd_delete(pid_timerMap[pid]);
+	}
+}
+
+/************************************************************************
+ * Public Functions
+ ************************************************************************/
+
+/************************************************************************
+ * Name: pm_timedStay
+ *
+ * Description:
+ *   This function locks PM transition for a specific duration.  
+ * 
+ * Parameters:
+ *   timer_interval - expected lock duration
+ *
+ * Return Value:
+ *   0 - success
+ *   -1 - error
+ *
+ ************************************************************************/
+
+int pm_timedStay(unsigned int timer_interval)
+{
+	int pid = getpid();
+
+	/* Check if there is already a wdog lock timer running for 
+	 * the process */
+	if (pid_timerMap[pid] != NULL) {
+		pmdbg("There is already a lock timer running for this process\n");
+		return PM_TIMER_FAIL;
+	}
+	
+	WDOG_ID wdog = wd_create();
+	pid_timerMap[pid] = wdog;
+
+	/* Lock the pm transition and Start the wdog timer */
+	pm_stay(PM_IDLE_DOMAIN, PM_NORMAL);
```

Comment:
wa can call pm_timedStay for other pm state as pm_idle, pm_standby.

Please add argument state and domain.

Reply:
let's seperate pm_timedStay_cancel

we can apply about pm_timedStay_cancel().
Please remove in this PR and raise new PR